internal/credspropagation/openstack.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2024
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package credspropagation
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hmc "github.com/Mirantis/hmc/api/v1alpha1"
+)
+
+func PropagateOpenStackSecrets(ctx context.Context, cfg *PropagationCfg) error {
+	openstackManagedCluster := &hmc.ManagedCluster{}
+	if err := cfg.Client.Get(ctx, client.ObjectKey{
+		Name:      cfg.ManagedCluster.Name,
+		Namespace: cfg.ManagedCluster.Namespace,
+	}, openstackManagedCluster); err != nil {
+		return fmt.Errorf("failed to get ManagedCluster %s: %w", cfg.ManagedCluster.Name, err)
+	}
+
+	openstackCredential := &hmc.Credential{}
+	if err := cfg.Client.Get(ctx, client.ObjectKey{
+		Name:      openstackManagedCluster.Spec.Credential,
+		Namespace: openstackManagedCluster.Namespace,
+	}, openstackCredential); err != nil {
+		return fmt.Errorf("failed to get OpenStackCredential %s: %w", cfg.ManagedCluster.Spec.Credential, err)
+	}
+
+	// Fetch the secret containing OpenStack credentials
+	openstackSecret := &corev1.Secret{}
+	openstackSecretName := openstackCredential.Spec.IdentityRef.Name
+	openstackSecretNamespace := openstackCredential.Spec.IdentityRef.Namespace
+	if err := cfg.Client.Get(ctx, client.ObjectKey{
+		Name:      openstackSecretName,
+		Namespace: openstackSecretNamespace,
+	}, openstackSecret); err != nil {
+		return fmt.Errorf("failed to get OpenStack secret %s: %w", openstackSecretName, err)
+	}
+
+	// Generate CCM secret
+	ccmSecret := generateOpenStackCCMSecret(openstackSecret)
+
+	// Apply CCM config
+	if err := applyCCMConfigs(ctx, cfg.KubeconfSecret, ccmSecret); err != nil {
+		return fmt.Errorf("failed to apply OpenStack CCM secret: %w", err)
+	}
+
+	return nil
+}
+
+func generateOpenStackCCMSecret(openstackSecret *corev1.Secret) *corev1.Secret {
```

Comment:
You should create the secret with OpenStack cloud configuration following the specific format. It's not the same as the secret with credentials in the management cluster (the format: https://github.com/kubernetes/cloud-provider-openstack/blob/master/manifests/controller-manager/cloud-config).

Related thing: credentials should not be exposed, so instead of passing the credentials as chart values you should pass the name of this secret to the CCM and CSI values: https://github.com/Mirantis/hmc/pull/696/files#diff-c5afee6dd916ada2ea759da67744b5845149af9a6532b3891574865168c2c44e
See: https://github.com/kubernetes/cloud-provider-openstack/blob/master/charts/cinder-csi-plugin/values.yaml#L162

Similarly, it's done for the azure and the vsphere providers. You can use it as a reference.

Reply:
@eromanova Yes, it already follows the same format. And based on the other comments, I have removed passing values to ccm and csi and am passing just the secret. So, this should ideally work. 

@a13x5 Yes, I defined a function here with the intention of doing the INI format thing. But later realized I don't have to do that. I agree that we don't need this function. Modified.


I will anyways test this out and make sure everything works.